### PR TITLE
Implement AbstractDataReader and RecordIODataReader

### PR DESCRIPTION
**Pull request by terrytangyuan**
Part of #1105. See design doc in https://github.com/sql-machine-learning/elasticdl/pull/1159. 

This includes the following:
* Add `AbstractDataReader` abstract class.
* Implement `RecordIODataReader` to support RecordIO data reading.
* Update `TaskDataService` and `master.main` to support `RecordIODataReader`.
* Rename `recordio_dataset()` to `create_dataset_from_tasks()` so it can be reused across different implementations of `AbstractDataReader`.

### CODE REVIEW

File: elasticdl/python/common/data_reader.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                else:
+                    break
+
+    def create_shards(self, mode):
+        if mode == Mode.TRAINING:
+            data_dir = self._kwargs["training_data_dir"]
+        elif mode == Mode.EVALUATION:
+            data_dir = self._kwargs["evaluation_data_dir"]
+        else:
+            data_dir = self._kwargs["prediction_data_dir"]
```
Comment (ywskycn) [inline elasticdl/python/common/data_reader.py]: Possible move this logic outside of the reader itself? The reader should just take a directory and create all shards.

Comment (terrytangyuan) [reply elasticdl/python/common/data_reader.py]: Good idea. Done!

**Review by mhaoli — APPROVED** [review]
LGTM

File: doc/custom_data_source_support_design.md [hunk 2]
```diff
@@ -79,13 +77,16 @@ class CustomDataReader(AbstractDataReader):
...
 
-    def create_shards(self, mode):
-        if mode == Mode.TRAINING:
-            return training_shards
-        elif mode == Mode.EVALUATION:
-            return evaluation_shards
-        else:
-            return prediction_shards
+    def create_shards(self):
+        data_dir = self._kwargs["data_dir"]
```
Comment (chunyang-wen) [inline doc/custom_data_source_support_design.md:84]: Check the required fields in the constructor such as checking `data_dir`?

File: elasticdl/python/worker/task_data_service.py [hunk 3]
```diff
@@ -123,14 +128,8 @@ def _gen(self):
...
-            ) as reader:
-                while True:
-                    record = reader.record()
-                    if record:
-                        yield record
-                    else:
-                        break
+            for data in self._data_reader.read_records(task):
+                if data:
+                    yield data
```
Comment (chunyang-wen) [inline elasticdl/python/worker/task_data_service.py:133]: Can we make `read_records` into a generator? If no data is returned, raise `StopIteration`?

File: doc/custom_data_source_support_design.md [hunk 4]
```diff
@@ -39,7 +39,7 @@ The interface would look like the following:
 
 ```python
 class AbstractDataReader(object):
-    def __init__(self, *kwargs):
+    def __init__(self, **kwargs):
```
Comment (chunyang-wen) [inline doc/custom_data_source_support_design.md:45]: This document is not complied with the code, such as `ABC`.

File: elasticdl/python/worker/task_data_service.py (hunk 3 above)
Comment (terrytangyuan) [reply elasticdl/python/worker/task_data_service.py:133]: Good idea. I'll improve this later.

**Review by chunyang-wen — APPROVED** [review]
TODO
+ make `read_records` return a generator object.

**Pull request merged**